ledger-app-fio/src/signTransaction.c

```diff
@@ -0,0 +1,796 @@
+
+		ctx->expiration = u4be_read(wireData->expiration);
+		sha_256_append(&ctx->hashContext, (uint8_t *)&ctx->expiration, sizeof(ctx->expiration));
+
+		ctx->refBlockNum = u2be_read(wireData->refBlockNum);
+		sha_256_append(&ctx->hashContext, (uint8_t *)&ctx->refBlockNum, SIZEOF(wireData->refBlockNum));
+
+		ctx->refBlockPrefix = u4be_read(wireData->refBlockPrefix);
+		sha_256_append(&ctx->hashContext, (uint8_t *)&ctx->refBlockPrefix, sizeof(ctx->expiration));
+
+		uint8_t buf[4]; //max_net_usage_words, max_cpu_usage_ms, delay_sec, context_free_actions
+		explicit_bzero(buf, sizeof(buf)); //SIZEOF does no work for 4
+		sha_256_append(&ctx->hashContext, buf, sizeof(buf));
+	}
+
+	security_policy_t policy = policyForSignTxHeader();
+	TRACE("Policy: %d", (int) policy);
+	ENSURE_NOT_DENIED(policy);
+	{
+		// select UI steps
+		switch (policy) {
+#	define  CASE(POLICY, UI_STEP) case POLICY: {ctx->ui_step=UI_STEP; break;}
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_HEADER_STEP_RESPOND);
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	signTx_handleHeader_ui_runStep();
+}
+
+// ============================== ACTION HEADER ==============================
+
+enum {
+	HANDLE_ACTION_HEADER_STEP_SHOW_TYPE = 300,
+	HANDLE_ACTION_HEADER_STEP_RESPOND,
+	HANDLE_ACTION_HEADER_STEP_INVALID,
+};
+
+static void signTx_handleActionHeader_ui_runStep()
+{
+	TRACE("UI step %d", ctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = signTx_handleActionHeader_ui_runStep;
+
+	UI_STEP_BEGIN(ctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_ACTION_HEADER_STEP_SHOW_TYPE) {
+		switch(ctx->action_type) {
+#	define  CASE(ACTION, STRING) case ACTION: {ui_displayPaginatedText("Action:", STRING, this_fn); break;}
+			CASE(ACTION_TYPE_TRNSFIOPUBKY, "Transfer FIO tokens");
+#	undef   CASE
+		default:
+			THROW(ERR_NOT_IMPLEMENTED);
+		}
+	}
+
+	UI_STEP(HANDLE_ACTION_HEADER_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+		advanceStage();
+	}
+
+	UI_STEP_END(HANDLE_ACTION_HEADER_STEP_INVALID);
+}
+
+__noinline_due_to_stack__
+void signTx_handleActionHeaderAPDU(uint8_t p2, uint8_t* wireDataBuffer, size_t wireDataSize)
+{
+	TRACE_STACK_USAGE();
+	{
+		// sanity checks
+		CHECK_STAGE(SIGN_STAGE_ACTION_HEADER);
+
+		VALIDATE(p2 == P2_UNUSED, ERR_INVALID_REQUEST_PARAMETERS);
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+
+	{
+		// parse data
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		struct {
+			uint8_t contractAccountName[CONTRACT_ACCOUNT_NAME_LENGTH];
+		}* wireData = (void*) wireDataBuffer;
+
+		VALIDATE(SIZEOF(*wireData) == wireDataSize, ERR_INVALID_DATA);
+
+		uint8_t buf[1];
+		buf[0] = 1;
+		sha_256_append(&ctx->hashContext, buf, SIZEOF(buf)); //one action
+		sha_256_append(&ctx->hashContext, (uint8_t *) wireData, SIZEOF(*wireData));
+
+		ctx->action_type = getActionTypeByContractAccountName(ctx->network, wireData->contractAccountName,
+		                   CONTRACT_ACCOUNT_NAME_LENGTH);
```

Comment:
The formatting of such "long" funtion calls is different in Cardano app, I think.